ocaml/xapi/xapi_sr.ml

```diff
@@ -387,15 +387,15 @@ let update_vdis ~__context ~sr db_vdis vdi_infos =
       | loc, None, Some v -> Some v
       | _, _, _ -> None
     ) db_vdi_map scan_vdi_map in
-  let to_update = StringMap.merge (fun loc db scan -> match loc, db, scan with
-      | loc, Some (r, v), Some vi -> Some (r, v, vi)
-      | _, _, _ -> None
-    ) db_vdi_map scan_vdi_map in
 
   let find_vdi db_vdi_map loc =
     if StringMap.mem loc db_vdi_map
     then fst (StringMap.find loc db_vdi_map)
-    else Ref.null in
+    else
+      (* CA-254515: Also check for the snapshoted VDI in the database *)
+      try Db.VDI.get_by_uuid ~__context ~uuid:loc
```

Comment:
I wonder what will happen in the case of a VDI from an ISO SR? The `location` of those VDIs are their filenames, and not their UUIDs:
```
# xe vdi-list name-label=centos63_x86-64_xenrtinst.iso params=location,uuid
uuid ( RO)        : 4925632f-c011-4df6-b8e7-1e8b414d508d
    location ( RO): centos63_x86-64_xenrtinst.iso
```
Although it's also true that the ISO SR does not support `VDI.update`, which calls this function:
```
# xe vdi-update uuid=4925632f-c011-4df6-b8e7-1e8b414d508d
The SR backend does not support the operation (check the SR's allowed operations)
sr: dd420cd3-76f8-9f45-5d7a-95bb4fc8a451 (XenRT Linux ISOs)
```
These operations seem to be supported by this SR:
```
# xe sr-list name-label=XenRT\ Linux\ ISOs params=allowed-operations
allowed-operations (SRO)    : unplug; plug; PBD.create; PBD.destroy; VDI.clone; scan; VDI.create; VDI.destroy
```
Of these, `VDI.clone` calls this function indirectly through `Xapi_vdi.update_vdi_db`.

This is an edge case, but maybe it would be safer to look for the VDI based on its location field?
The `location` field of this VDI seems to match with its `vdi` field from SM:
```
# sm-cli sr-scan dd420cd3-76f8-9f45-5d7a-95bb4fc8a451 | grep centos64_x86-64
centos64_x86-64_xenrtinst.iso: {"vdi": "centos64_x86-64_xenrtinst.iso", "uuid": "0cd643c8-10ab-4b03-b140-fd7f922fb649", "content_id": "centos64_x86-64_xenrtinst.iso", "name_label": "centos64_x86-64_xenrtinst.iso", "name_description": "", "ty": "user", "metadata_of_pool": "OpaqueRef:NULL", "is_a_snapshot": false, "snapshot_time": "19700101T00:00:00Z", "snapshot_of": "", "read_only": true, "virtual_size": 4357423104, "physical_utilisation": 4357423104, "persistent": true, "sm_config": {}}
```